scripts/pages/zoning/zoning.ts

```diff
@@ -0,0 +1,922 @@
+
+	static addPointToList(i: number, point: number[]) {
+		const newRegionPoint = $.CreatePanel('Panel', this.panels.pointsList, `Point ${i}`);
+		newRegionPoint.LoadLayoutSnippet('region-point');
+		(newRegionPoint.FindChildTraverse('PointX') as TextEntry).text = point[0].toFixed(2);
+		(newRegionPoint.FindChildTraverse('PointY') as TextEntry).text = point[1].toFixed(2);
+		const deleteButton = newRegionPoint.FindChildTraverse('DeleteButton') as Panel;
+		deleteButton.SetPanelEvent('onactivate', () => {
+			ZoneMenu.deleteRegionPoint(newRegionPoint);
+		});
+	}
+
+	static deleteRegionPoint(point: Panel) {
+		if (!this.selectedZone || !this.selectedZone.zone) return;
+
+		const n = this.panels.pointsList.Children().indexOf(point);
+		const index = this.panels.regionSelect.GetSelected().GetAttributeInt('value', -1);
+		this.selectedZone.zone?.regions[index].points.splice(n, 1);
+		point.DeleteAsync(0);
+	}
+
+	static addRegionPoint(point) {
+		if (!this.selectedZone || !this.selectedZone.zone) return;
+
+		const index = this.panels.regionSelect.GetSelected().GetAttributeInt('value', -1);
+		if (index > -1) {
+			const i = this.selectedZone.zone.regions[index].points.length;
+			this.selectedZone.zone.regions[index].points.push([point.x, point.y]);
+			this.addPointToList(i, [point.x, point.y]);
+		}
+	}
+
+	static setRegionBottom() {
+		if (!this.selectedZone || !this.selectedZone.zone) return;
+
+		const regionIndex = this.panels.regionSelect.GetSelected().GetAttributeInt('value', -1);
+		const region = this.selectedZone.zone.regions[regionIndex];
+		const bottom = Number.parseFloat(this.panels.regionBottom.text);
+		region.bottom = Number.isNaN(bottom) ? 0 : bottom;
+	}
+
+	static setRegionHeight() {
+		if (!this.selectedZone || !this.selectedZone.zone) return;
+
+		const regionIndex = this.panels.regionSelect.GetSelected().GetAttributeInt('value', -1);
+		const region = this.selectedZone.zone.regions[regionIndex];
+		const height = Number.parseFloat(this.panels.regionHeight.text);
+		region.height = Number.isNaN(height) ? 0 : height;
+	}
+
+	static setRegionSafeHeight() {
+		if (!this.selectedZone || !this.selectedZone.zone) return;
+
+		const regionIndex = this.panels.regionSelect.GetSelected().GetAttributeInt('value', -1);
+		const region = this.selectedZone.zone.regions[regionIndex];
+		const height = Number.parseFloat(this.panels.regionSafeHeight.text);
+		region.regionSafeHeight = Number.isNaN(height) ? 0 : height;
+	}
+
+	static updateRegionTPDest() {
+		if (!this.selectedZone || !this.selectedZone.zone || !this.teleDestList) return;
+
+		const regionIndex = this.panels.regionSelect.GetSelected().GetAttributeInt('value', -1);
+		const teleDestIndex = this.panels.regionTPDest.GetSelected()?.GetAttributeInt('value', 0);
+		const region = this.selectedZone.zone.regions[regionIndex];
+		if (teleDestIndex === 0) {
+			// no teleport destination for this region
+			region.teleDestTargetname = '';
+			//@ts-expect-error property must be optional
```

Comment:
?
Wrap in `if (region.teleDestTargetname)` if needed, TS will narrow on that

Reply:
What do you mean by the //@ts-expect-error comment?

Reply:
maybe it's a strict mode thing, but I can't delete `teleDestPos` or `teleDestYaw` from `region` without an error. These are optional properties, but I couldn't define them as optional in the interface with either `?:` or declaring the type as a union with `undefined`. The c++ validator will either complain if either of those properties exists while `teleDestTargetname` is not the empty string, or use them as the tp destination if `teleDestTargetname` is the empty string.